### PR TITLE
Add conditions to camel-k CRs

### LINKED ISSUE #594: Add conditions to camel-k CRs
In order to improve visibility of what happened to camel-k CRs, we should add conditions to CRs, like: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions

### PR DESCRIPTION
**Pull request by lburgazzoli**
Fixes #594

I did a little bit of work to introduce `conditions` in camel-k CRs so now if you invoke 

    kamel describe integration hello

you'd get an output like:

```yaml
Name:                hello
Namespace:           myproject
Creation Timestamp:  Sun, 23 Jun 2019 11:46:00 +0200
Phase:               Running
Camel Version:       2.24.0
Kit:                 kit-bk7kh138gd7vch0m1d00
Image:               172.30.1.1:5000/myproject/camel-k-kit-bk7kh138gd7vch0m1d00:843534
Dependencies:
  camel:core
  runtime:jvm
Sources:
  Name       Language  Compression  Ref  Ref Key
  hello.xml  xml       false             
Conditions:
  Type                 Status  Reason                        Message
  PlatformAvailable    True    IntegrationPlatformAvailable  IntegrationPlatform (camel-k)
  KitAvailable         True    IntegrationKitAvailable       IntegrationKit (kit-bk7kh138gd7vch0m1d00)
  DeploymentAvailable  True    Deployment                    Deployment
  ExposureAvailable    True    Service                       Service
Traits:
  Route:
    Configuration:
      Enabled:  false
  Service:
    Configuration:
      Auto:     false
      Enabled:  true
```

which includes a number of `conditions` that represents the latest available observations of a integration state. 

The goal here is to provide more insight about the integration status i.e. the additional  resources associated to the integration that the operator has created (deployment vs knative service, service/route/ingress, etc) so user and tools do not need to explore the integration namespace to know what are the other resources related to the integration.

@astefanutti @nicolaferraro does this make sense ?

### CODE REVIEW

**Comment by lburgazzoli** [comment]
> Love it!
> 
> Two general comments:
> 
> * Currently, conditions are shared among CRs, so it is possible to mix them, I wonder whether there should be dedicated conditions types per CR

Yes, that’s what I’m working on. It is a little bit of a pain as we basically need to duplicate all the code but that’t the way to *go* :)

> * It seems there is a larger need to report the owned resources and I wonder to what extend it generalises the proposed `DeploymentAvailable` and `ExposureAvailable` conditions

Yes, the example here is just a placeholder to show how it will look but the final goal is to report any owned resource

**Comment by lburgazzoli** [comment]
So I've added a little bit more logic and now if you describe an integration while it is building a kit you'll see something like:

```
Name:                routes-rest
Namespace:           myproject
Creation Timestamp:  Mon, 08 Jul 2019 13:34:02 +0200
Phase:               Building Kit
Camel Version:       3.0.0-M3
Kit:                 kit-bkhikar8gd7uoo13f47g
Image:               
Version:             1.0.0-M1-SNAPSHOT
Dependencies:
  camel:core
  camel:log
  camel:timer
  camel:undertow
  runtime:jvm
Sources:
  Name            Language  Compression  Ref  Ref Key
  routes-rest.js  js        false             
Conditions:
  Type                          Status  Reason                        Message
  IntegrationPlatformAvailable  True    IntegrationPlatformAvailable  camel-k
  IntegrationKitAvailable       False   IntegrationKitAvailable       
```

Once the kit is ready, then the `IntegrationKitAvailable` condition become true and  other info such as services an routes are reported:

```
Name:                routes-rest
Namespace:           myproject
Creation Timestamp:  Mon, 08 Jul 2019 13:34:02 +0200
Phase:               Running
Camel Version:       3.0.0-M3
Kit:                 kit-bkhikar8gd7uoo13f47g
Image:               172.30.1.1:5000/myproject/camel-k-kit-bkhikar8gd7uoo13f47g:1281056
Version:             1.0.0-M1-SNAPSHOT
Dependencies:
  camel:core
  camel:log
  camel:timer
  camel:undertow
  runtime:jvm
Sources:
  Name            Language  Compression  Ref  Ref Key
  routes-rest.js  js        false             
Conditions:
  Type                          Status  Reason                        Message
  IntegrationPlatformAvailable  True    IntegrationPlatformAvailable  camel-k
  IntegrationKitAvailable       True    IntegrationKitAvailable       kit-bkhikar8gd7uoo13f47g
  DeploymentAvailable           True    DeploymentAvailable           routes-rest
  ExposureAvailable             True    RouteAvailable                routes-rest -> routes-rest(http)
  ServiceAvailable              True    ServiceAvailable              routes-rest(http/80) -> routes-rest(http/8080)
```

If you run the same integration but with the `service` trait disabled, you should see something like:

```
ame:                routes-rest
Namespace:           myproject
Creation Timestamp:  Mon, 08 Jul 2019 13:37:12 +0200
Phase:               Running
Camel Version:       3.0.0-M3
Kit:                 kit-bkhikar8gd7uoo13f47g
Image:               172.30.1.1:5000/myproject/camel-k-kit-bkhikar8gd7uoo13f47g:1281056
Version:             1.0.0-M1-SNAPSHOT
Dependencies:
  camel:core
  camel:log
  camel:timer
  camel:undertow
  runtime:jvm
Sources:
  Name            Language  Compression  Ref  Ref Key
  routes-rest.js  js        false             
Conditions:
  Type                          Status  Reason                        Message
  IntegrationPlatformAvailable  True    IntegrationPlatformAvailable  camel-k
  ServiceAvailable              False   ServiceNotAvailable           explicitly disabled
  IntegrationKitAvailable       True    IntegrationKitAvailable       kit-bkhikar8gd7uoo13f47g
  DeploymentAvailable           True    DeploymentAvailable           routes-rest
  ExposureAvailable             False   RouteNotAvailable             no target service found
```

**Comment by lburgazzoli** [comment]
@astefanutti @nicolaferraro we still miss to include additional information in the conditions but didn't want to make the PR too big, can you please review ?

File: pkg/controller/integration/integration_controller.go [hunk 1]
```diff
@@ -78,6 +79,14 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		UpdateFunc: func(e event.UpdateEvent) bool {
 			oldIntegration := e.ObjectOld.(*v1alpha1.Integration)
 			newIntegration := e.ObjectNew.(*v1alpha1.Integration)
+
```
Comment (astefanutti) [inline pkg/controller/integration/integration_controller.go]: My understanding is that this statement is already logged in the `Reconcile` method.

Comment (lburgazzoli) [reply pkg/controller/integration/integration_controller.go]: oh yeah, that's a leftover from the rebase, will fix it

**Comment by astefanutti** [comment]
I wonder if it'd be possible to have a base type for the conditions that would enable factorising some of the helper methods as well.

**Comment by lburgazzoli** [comment]
yeah, who needs generics ? :P 
I can have a look but haven't found any concrete example so fa on kubernetes for a best practice

**Comment by astefanutti** [comment]
Let's wait for go to have generics 😄

File: pkg/controller/integration/integration_controller.go (hunk 1 above)
Comment (lburgazzoli) [reply pkg/controller/integration/integration_controller.go]: done

**Comment by lburgazzoli** [comment]
@asfsecurity can this be merged ?

**Comment by astefanutti** [comment]
Good for me 😉

**Pull request merged**